grpc-protocol/src/main/java/com/linecorp/armeria/common/grpc/protocol/ArmeriaMessageFramer.java

```diff
@@ -137,10 +142,23 @@ public HttpData writePayload(ByteBuf message, boolean webTrailers) {
             } else {
                 buf = writeUncompressed(message, webTrailers);
             }
+
+            final ByteBuf maybeEncodedBuf;
+            if (encodeBase64) {
+                final ByteBuffer base64Encoded = Base64.getEncoder().encode(buf.nioBuffer());
+                buf.release();
+                if (maxOutboundMessageSize >= 0 && base64Encoded.remaining() > maxOutboundMessageSize) {
+                    throw newMessageTooLargeException(base64Encoded.remaining());
+                }
+                maybeEncodedBuf = Unpooled.wrappedBuffer(base64Encoded);
+            } else {
+                maybeEncodedBuf = buf;
+            }
+
             if (webTrailers) {
-                return HttpData.wrap(buf).withEndOfStream();
+                return HttpData.wrap(maybeEncodedBuf).withEndOfStream();
             }
-            return HttpData.wrap(buf);
+            return HttpData.wrap(maybeEncodedBuf);
```

Comment:
Could be simplified into `return HttpData.wrap(maybeEncodedBuf).withEndOfStream(webTrailers);`

Reply:
Fixed. 😉